proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,302 @@
+cannot be an intangible type, there is only one exception where the register keyword
+can legally be applied to the variable. The only valid register type for such a 
+numeric variable type is 'c'. The register keyword in this statement:
+`float f : register(c0)`
+isn't binding a resource, rather it is specifying a constant register binding offset
+within the $Globals cbuffer, which is legacy behavior from DX9. Because this specific 
+register overload is only applicable if the developer wants to make modifications to the
+$Globals buffer, it is not useful when the overload is not used inside the $Globals
+context. That is, if this register overload appears within a `cbuffer {...}` or 
+`tbuffer {...}` scope, the register number will be ignored. In this case, 
+`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, and will be treated as an error by default.
+All other register types applied to such a type will emit `err_hlsl_unsupported_register_prefix"`, except for 'b' or 'i'.
+If the register type is 'b' or 'i', then `warn_hlsl_deprecated_register_type_b` or `warn_hlsl_deprecated_register_type_i` 
+will be emitted respectively, and treated as errors by default.
+
+Below are some examples:
+
+| Code | Diagnostic |
+|-|-|
+| `float f : register(t0)` | "error: unsupported resource register binding 't' on variable of type 'float'" |
+| `float f : register(c0)` | Valid, no errors
+| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." |
+| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used." |
+| `float f : register(x0)` | "error: invalid resource class specifier 'x' used; expected 't', 'u', 'b', or 's'" |
+| `cbuffer g_cbuffer { float f : register(c2); }` | "warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead" |
+| `RWBuffer<float> f : register(c3);`| "error: UAV type 'RWBuffer<float>' requires register type 'u', but register type 'c' was used." |
+
+
+## Detailed design
+
+### Analysis 
+
+All the compiler has to work with is a Decl object and the context the decl appears in.
+From this information, the first goal is to set a specific set of flags that can fully 
+inform the compiler of the right diagnostic to emit. The first group of flags are the 
+decl type flags, which are either `basic`, `resource`, `udt`, or `other`. `basic` refers 
+to a numeric variable type. `other` refers to built-in HLSL object types (intangible types) 
+that are not resources, and thus cannot be placed into a cbuffer or bound as a resource.
+
+These flags are all mutually exclusive, and are represented in the code with an enum.
+The `udt` flag has an associated flag, `contains_numeric`, which will be set if the
+UDT contains a numeric type.
+
+The next group of flags are the resource class flags,
+which are `srv`, `uav`, `cbv`, or `sampler`.

```

Comment:
I see references to 'other' flag below.  It doesn't appear in this list of supported flags..  Should it be here?  With a definition of what it represents?

Reply:
`other` is in the "decl type" group of flags, and there's a short explanation in the first paragraph describing what it represents.